### ISSUE TITLE
Block USB keyboard in linux

### ISSUE THREAD

**Issue by atarijookie** [atarijookie/ce-atari#134]
I swear to flying spaghetti monster that I've created this one, but I can't find it. If it's here somewhere, we can link them and close this one.

When using USB keyboard through CE, the input not only goes to ST, but also to linux console on RPi, which is auto-logged in in Raspbian.

Things to do:
1) disable auto login, so user won't be logged in automatically (seen through HDMI)
2) disable linux console input from USB, add some switching between linux console and Atari IKDB

@mikrosk - do you have any idea how to solve 2), or at least do the preparation for switching that keyboard, like you're currently doing for your IKBD fake joysticks switching?

**Comment by mikrosk** [comment]
If Linux ce_conf works the same way the Atari version does (i.e. sending 'alive' pings to the CE), you'll get this for free in #73. I have it implemented since a couple of weeks ago.

**Comment by mikrosk** [comment]
But I'm not sure what are you trying to prevent? So if I have a text editor open on Atari and I run CE_CONF in Linux, you don't want to see moving down in the text editor if I'm moving down in the menu?

**Comment by atarijookie** [comment]
> But I'm not sure what are you trying to prevent?

Currently:
Open a text editor / e-mail editor / game, and you - for some magical reasons - want to write a readme.txt / some e-mail / your geek name containing 'rm -rf /' and bam - you just unknowingly deleted your RPi linux files, while you were typing something on Atari through USB keyboard. The same goes for 'shutdown', 'reboot' and so on.

New behavior:
Until you press the magical combination, you can type on Atari 'rm -rf /' and not delete your whole linux in CE.

**Comment by atarijookie** [comment]
PS: replying to this was difficult:
[https://www.youtube.com/watch?v=_TwCIxy1ExE](https://www.youtube.com/watch?v=_TwCIxy1ExE)

**Comment by tin-nl** [comment]
Incidently I just observed a similar issue: I rebooted my CE during some tests by accident just by pressing CTRL-ALT-DEL via USB during boot :)
So if say, someone is waiting eagerly for CE to boot (*cough* @mikrosk *cough* ;-) ), he might reboot the CE while rebooting his ST via keyboard...

And: if ce_main is needed to activate console input via keyboard, troubleshooting a device might get problematic...

**Comment by atarijookie** [comment]
> I rebooted my CE during some tests by accident just by pressing CTRL-ALT-DEL via USB during boot :)

Does that Raspbian / Yocto react to Ctrl-Alt-Del? I didn't know that :)

> And: if ce_main is needed to activate console input via keyboard, troubleshooting a device might get problematic...

Only for the troubleshooting on TV + USB keyboard - the ssh to device shouldn't be affected. If ce_main would be running, then even switching the for linux on and off shouldn't be a problem, the only case would be a problem when the ce_main refuses to run for some reason (or keeps terminating, when HW doesn't respond).

**Comment by tin-nl** [comment]
> Does that Raspbian / Yocto react to Ctrl-Alt-Del? I didn't know that :)

Yeah, surprising news to me, too :) Yocto isn't affected by this.

> Only for the troubleshooting on TV + USB keyboard - the ssh to device shouldn't be affected.

Yeah, I know. But that's (HDMI/Kbd) exactly the thing that people w/o network or with borked up CE/network config might need to unbrick their device.
No idea for a better solution atm, though :/

**Comment by mikrosk** [comment]
Hehe, this is a fascinating problem, had no idea at all. I use Atari keyboard for my IKBD stuff (with occasional testing for its USB counterpart but never for typing), wow.

Well, if you says that you're actually typing in _command line_, the problem lies elsewhere, not in CE_CONF. You must first disable autologin (as you say) plus you must somehow prevent even having login prompt at all. And frankly, I have no clue how to achieve this.

**Comment by mikrosk** [comment]
After some thinking -- we need that autologin because of "F8" functionality (i.e. switching to bash from Atari's CE_CONF)? If so, this is perhaps fixable with asking for login/password instead of skipping directly to command line.

But about disabling login prompt after boot... no clue. We could perhaps autostart an app in the foreground (!) which would eat all input (which could be possible to terminate via CTRL+C on HDMI) ? (with this hack we perhaps wouldn't need to disable autologin, just open another terminal on F8).

**Comment by atarijookie** [comment]
> We need that autologin because of "F8" functionality (i.e. switching to bash from Atari's CE_CONF)?

NO, that is done through fork() and exec(), that HDMI console is not needed, unless you want to try to resolve some issue with CE on TV.

I wouldn't disable login prompt, I for now I would just hope nobody writes 'root[enter]ce[enter]' on USB keyboard while doing something on Atari ;)

Btw., if I remember correctly, the typical login program could be replaced with something else, which would just wait for a special key combination (e.g. Ctrl - Shift - F8) and that one would exec then the typical login prompt... 

But I guess leaving it at login prompt would be probably just fine most of the time.

**Comment by mikrosk** [comment]
What about that idea of yours, the magic hotkey? Can't we replace /bin/login with a program which listens for LCTRL+LALT+RCTRL+RALT+DEL (impossible to do on Atari keyboard) and then show the login prompt? And if you reboot/power cycle the device, you're again in the fake login.

**Comment by miniupnp** [comment]
F11 or F12 are good magic hotkeys too ;)

**Comment by mikrosk** [comment]
> F11 or F12 are good magic hotkeys too ;)

Not really, they are mapped to UNDO/HELP (not in master but in my branch for #73)

**Comment by miniupnp** [comment]
why not map Page Up / Page Down to Undo/help ? It is what hatari does

**Comment by mikrosk** [comment]
Historically I remember it from Eiffel that way -- PU/PD was mapped to SHIFT+UP/SHIFT+DOWN (something I plan to do after #73, it requires multi mapping from one code to many).

**Comment by atarijookie** [comment]
> What about that idea of yours, the magic hotkey? Can't we replace /bin/login...

Yeah, that could be done, too, but it seems like too much work, and might behave differently on Yocto and Raspbian, while no one (yet) has complained about this not-blocked USB keyboard. So I would just go with no-autologin and no-ctrl-alt-del for now, and later the rest... 

I'm still having trouble with #100

**Comment by mikrosk** [comment]
@atarijookie: sounds reasonable although now I will be much more cautious when using USB keyboards. :D

**Comment by atarijookie** [comment]
Disabling ctrl-alt-del:
```
sudo rm /lib/systemd/system/ctrl-alt-del.target
sudo ln -s /dev/null /lib/systemd/system/ctrl-alt-del.target
sudo systemctl daemon-reload
```

Disable auto-login:
```
ln -fs /lib/systemd/system/getty@.service
 /etc/systemd/system/getty.target.wants/getty@tty1.service
```

**Comment by miniupnp** [comment]
I'm just thinking, is there a way for the ce_main_app to grab exclusive access of the usb keyboard ?
The magic key would then be to release that exclusive access :)

**Comment by atarijookie** [comment]
> I'm just thinking, is there a way for the ce_main_app to grab exclusive access of the usb keyboard ?

I'm not sure, in theory yes, but didn't see / try that... But good idea :+1:

**Comment by miniupnp** [comment]
http://stackoverflow.com/questions/1698423/how-can-you-take-ownership-of-a-hid-device/1698686#1698686

ioctl(fd, EVIOCGRAB, (void*)1); // grab exclusive 
..
ioctl(fd, EVIOCGRAB, (void*)0); // release

**Comment by atarijookie** [comment]
Wow, nice, thanks ;) If @mikrosk  adds another magical keyboard shortcut, adding this one would be easy ;) Or even MiKRO could then add it while he's doing his IKBD enhancements...

**Comment by tin-nl** [comment]
Disabling C-A-D and grab all input via ce_main sounds good! There should probably be some text on the console above the login explaining the keystroke to unlock the keyboard.

**Comment by atarijookie** [comment]
I've added disabling Ctrl-Alt-Del and disabling auto-login into shellscripts/copynewscripts.sh

> There should probably be some text on the console above the login explaining the keystroke to unlock the keyboard.

Agree, I would just add it into **/etc/motd** which currently shows 'Raspbian Jessie running on CosmosEx device.'

@mikrosk - Will you do this improvement while you're playing with magical key strokes, or should I do it?

**Comment by mikrosk** [comment]
Wait, how do you type CTRL+ALT+DEL with **disabled  HID device**? :D

**Comment by mikrosk** [comment]
EDIT: deleted, I'm an idiot. ;)

Sure, I can take a look. So just to verify we understand each other:
- CE would boot with autologin disabled but with the login prompt
- if you try to type in root + ce, you can't because a small tool has grabbed the HID device
- this tool listens to CTRL+ALT+DEL (so we actually don't need that `ctrl-alt-del.target` stuff)
- if grabbed, the tool terminates
- for re-enabling the grab one would need to reboot the CE

Correct?

**Comment by atarijookie** [comment]
> Correct?

No :)

- CE would boot with autologin disabled (done in the script, applied with next update), with login prompt
- Ctrl+Alt+Del sequence for reboot will be disabled (done in the script, applied with next update)
- the ce_main_app is running in the background, and it did grab the excluside access to USB keyboard (when keyboard available)
- you have one magic key combination for enabling / disabling keyboard emulated joysticks
- so you add another magic key combination, which will toggle between excluside access to USB keyboard and anyone-receives-event-from-USB-keyboard state
- when user wants to login with the HDMI + USB keyboard, then he presses the newly added magic combination, and ce_main_app running in the background unlocks his keyboard, allowing him to type username and password
- I assume that when you terminate ce_main_app (even with kill -9), system will automatically release that USB keyboard grab

**Comment by mikrosk** [comment]
Ah, right. I've mistaken CE_CONF for ce_main_app, assuming that it can't run in background the whole time (I'm tired :-P).

It's a good concept, this could really work nicely.

EDIT: that being said, I fail to see why do we need these:

- CE would boot with autologin disabled (done in the script, applied with next update), with login prompt
- Ctrl+Alt+Del sequence for reboot will be disabled (done in the script, applied with next update)

at all?

**Comment by tin-nl** [comment]
> Ctrl+Alt+Del sequence for reboot will be disabled (done in the script, applied with next update)

I think that even in the above scenario the user _could_ reboot the CE by accident while ce_main isn't running - we don't know yet if ce_main starts early enough to prevent that.

That accidental reboot probably happened to to users already (but probably unbeknown to them).

**Comment by mikrosk** [comment]
> I think that even in the above scenario the user could reboot the CE by accident while ce_main isn't running - we don't know yet if ce_main starts early enough to prevent that.

So the assumption is that the systemd flag kicks in earlier?

And why do we need to disable autologin?

**Comment by atarijookie** [comment]
>I think that even in the above scenario the user could reboot the CE by accident while ce_main isn't running - we don't know yet if ce_main starts early enough to prevent that.

No no, by executing this (in the script):
```
ln -fs /dev/null /lib/systemd/system/ctrl-alt-del.target
```

It will disable ctrl-alt-del reboot forever (until changed back), so after that even when linux is booting and ce_main_app is not running, it will not reboot RPi. 

> And why do we need to disable autologin?

In case you still would be typing in console, while your keyboard grabbing feature is still not available ;)

**Comment by mikrosk** [comment]
> while your keyboard grabbing feature is still not available ;)

Why it wouldn't be available?

**Comment by miniupnp** [comment]
Le 10.01.2017 à 10:31, Miro Kropacek a écrit :
>
>     while your keyboard grabbing feature is still not available ;)
>
> Why it wouldn't be available?
>
until the ce_main_app is launched I guess...

**Comment by atarijookie** [comment]
> while your keyboard grabbing feature is still not available ;)
> Why it wouldn't be available?

Because you haven't done it, commited it and pushed it to master branch? :)

> until the ce_main_app is launched I guess...

No, it's a linux system setting - once you set it, it's there, independent of ce_main_app running or not.

**Comment by tin-nl** [comment]
> It will disable ctrl-alt-del reboot forever (until changed back), so after that even when linux is booting and ce_main_app is not running, it will not reboot RPi.

@atarijookie: I know, I was just trying to paint a picture that might happen if we don't do this :)
Edit: In that comment I was replying to Mikro, but did that poorly, sorry.

**Comment by miniupnp** [comment]
the ce_main_app will be grabbing the keyboard. But until it is runing, the USB keyboard events still goes to the system console. That's why it is needed to disable CTRL-ALT-DEL and disable autologin : 
the user can be logged a few seconds before ce_main_app is launched. Except if you can make sure autologin happens AFTER ce_main_app is able to grab USB keyboard exclusivity

**Comment by miniupnp** [comment]
I can code USB Keyboard grabbing, but I have no HDMI screen to test ;)

**Comment by mikrosk** [comment]
@miniupnp I don't have HDMI either. :-P IMO that autologin off is not needed (yes jookie, after I commit it ;)). My memory is vague here but from what I remember login happens at a certain init level so we just have to make sure to run ce_main_app on a lower one (what is the case right now, I believe).

**Comment by tin-nl** [comment]
@mikrosk: on raspbian ce_main starts directly after mountall.
@miniupnp: I can test that (via dvi adapter).

**Comment by miniupnp** [comment]
@tin-nl : could you test my branch https://github.com/miniupnp/ce-atari/tree/ikbd ?
only grabbing is implemented :)

**Comment by tin-nl** [comment]
Cool, sure :) I'll test yocto & raspbian.

**Comment by tin-nl** [comment]
@miniupnp; works fine on Raspbian - takes keyboard as soon as ce_main runs (faaaar before login), no input possible, input is routed to ST nicely. I can finally reboot my STe via keyboard w/o reboot the CE by accident ;-)

Could you shoot me a yocto binary? I can't cross compile to yocto ATM as it seems...

**Comment by miniupnp** [comment]
@tin-nl http://nanard.free.fr/CosmosEx/cosmosex_yocto_20170110.tar.gz

**Comment by tin-nl** [comment]
@miniupnp: Thanks a lot - works on yocto, too :) 
Couldn't check IKBD forwarding aka if ce_main still gets events (I see no reason why it shouldn't).

I also can confirm that ce_main releases USB on SIGINT/KILL (ce_stop.sh). Dunno what happens if it's crashing, though.

**Comment by miniupnp** [comment]
@tin-nl well it would be a linux kernel bug if the process keep exclusive access after exit, whatever the reason of the exit (crash, SIGTERM/SIGKILL/etc)

**Comment by miniupnp** [comment]
@atarijookie / @mikrosk : should I merge that code right away, or wait for the "magic key" to be implemented for release ?
I can implement it if you tell me which magic key combination you want

**Comment by mikrosk** [comment]
@miniupnp: go ahead, I wont be able to do it earlier than during the weekend and it's possible the last pending tasks for #73 will eat my all time.

**Comment by atarijookie** [comment]
I don't have a preference on Magic Key Combination, so maybe MiKRO should tell ;)

**Comment by miniupnp** [comment]
I've implemented it with LCTRL+LALT+RCTRL+RALT
if you're OK, I merge

**Comment by atarijookie** [comment]
I'm fine with it, and how about @mikrosk ?

**Comment by mikrosk** [comment]
@mikrosk is fine with that as well. So what's left to be implemented, if anything?

**Comment by miniupnp** [comment]
left to be implemented : 
update motd / documentation so user are aware of LCTRL+LALT+RCTRL+RALT

**Comment by mikrosk** [comment]
Ah, right, because we don't have an CE_CONF entry about that. Btw, help related, I've been thinking that every screen could have a help section. I.e. you press F11/HELP and you'd see a short description what's what, similar to HDDriver.

**Comment by miniupnp** [comment]
it has nothing to do with CE_CONF. A user just plugging the RPi to HDMI output + USB keyboard have to be informed on how to activate the keyboard !

**Comment by mikrosk** [comment]
@miniupnp: I'm aware of that. :) That's what I meant by "because we don't have an CE_CONF entry about that". Perhaps I should separate my thoughts because I just create chaos. The help key idea is unrelated to this particular issue.

**Comment by atarijookie** [comment]
> left to be implemented :
> update motd / documentation so user are aware of LCTRL+LALT+RCTRL+RALT

I will add that in the shell script to modify motd, possibly create a web help page for that...

**Comment by miniupnp** [comment]
@atarijookie you may need to edit /etc/issue instead, it is shown BEFORE login.
http://unix.stackexchange.com/questions/16255/how-can-i-change-whats-displayed-at-a-login-shell

**Comment by atarijookie** [comment]
You're right, motd is shown after login :+1:

**Comment by mikrosk** [comment]
Guys, didn't we forget about one scenario? Now I have used the magic combo to unlock the keyboard so I can type in shell, shouldn't Atari keyboard emulation be disabled until I lock it again? Because Atari is mirroring my shell commands, quite odd. :)
